### DISCUSSION TITLE
A2DP source and BLE simultaneously

### DISCUSSION THREAD

**Discussion by cyberFIRM** [pschatzmann/arduino-audio-tools#1228] [Q&A] [answered]
Has anyone ever tried to implement BLE beacon and A2DP source in single sketch?
### _Application example:_
I have a positioning proximity beacon. I want ESP32 to check if someone is within certain distance (lets say 1 meter), ESP32 plays an audio on the Bluetooth speaker.

I have tried to implement this application, but gets the following error when connecting to Bluetooth speaker:
```
[I] AudioA2DP.h : 126 - Connecting to E600Pro
[I] AudioA2DP.h : 137 - Starting a2dp_source...
ESP_ERROR_CHECK failed: esp_err_t 0x103 (ESP_ERR_INVALID_STATE) at 0x400944a0
file: ".pio/libdeps/esp32dev/ESP32-A2DP/src/BluetoothA2DPSource.cpp" line 203
func: virtual void BluetoothA2DPSource::start_raw(std::vector<const char*>, music_data_cb_t)
expression: esp_bt_controller_mem_release(ESP_BT_MODE_BLE)

abort() was called at PC 0x400944a3 on core 1
```

I am using player-sdfat-a2dp example as a reference.

### **Things I tried:**

1. Deinit BLE before connecting to Bluetooth speaker.
2. Use of Tasks in the code to have complete control of BLE task and A2DP task
3. mem release before init of BT and deinit of BLE
4. Identified the problem that esp_bt_controller_enable [can't be called twice using different configs](https://github.com/espressif/esp-idf/blob/165e3edd4949586dc90fa682f907d971fee488fc/components/bt/include/esp_bt.h#L161) and both libraries call it when they start

For both work together, they should call esp_bt_controller_enable(ESP_BT_MODE_BTDM) once, and not the BLE lib call esp_bt_controller_enable(ESP_BT_MODE_BLE) and BluetoohSerial call esp_bt_controller_enable(ESP_BT_MODE_CLASSIC_BT)
Still was not successful in this.

Any ideas?

If I run BLE beacon part or A2DP separately it works fine, but running both simultaneously does not work.

Thanks in advance

**Comment by pschatzmann** [comment, marked as answer]
Not sure if this helps, but I [committed a correction](https://github.com/pschatzmann/ESP32-A2DP/commit/c80c25034bc4b7de4296bda09c116da9ecf70cb2) to the [ESP-A2DP](https://github.com/pschatzmann/ESP32-A2DP) project:
- the ble memory does not get released any more by default
- you can define the bt_mode by calling the set_default_bt_mode() method

**Reply by cyberFIRM** [reply]
Thank you for replying back. I will try that out and will let you know.

**Reply by cyberFIRM** [reply]
Hi @pschatzmann ,

Thank you for the commit. I was able to connect to the bluetooth speaker but I cannot hear any sound. In the SD card, there is only one file name 1.mp3
Debug logs are below:
```
00:35:47.302 -> [W] AudioA2DP.h : 277 - ==> state: Connecting
00:35:47.712 -> [D] AudioA2DP.h : 148 - waiting for connection
00:35:48.703 -> [D] AudioA2DP.h : 148 - waiting for connection
00:35:49.602 -> [D] AudioA2DP.h : 272 - static void audio_tools::A2DPStream::a2dp_state_callback(esp_a2d_connection_state_t, void*)
00:35:49.603 -> [W] AudioA2DP.h : 277 - ==> state: Connected
00:35:49.708 -> [I] AudioA2DP.h : 151 - a2dp_source is connected...
00:35:49.708 -> test 4 !!
00:35:49.708 -> [I] AudioPlayer.h : 344 - setVolume(0.100000)
00:35:49.708 -> [I] VolumeStream.h : 198 - setVolume: 0.100000
00:35:49.708 -> [I] VolumeStream.h : 198 - setVolume: 0.100000
00:35:49.708 -> test 5 !!
00:35:49.708 -> [D] AudioPlayer.h : 154 - virtual bool audio_tools::AudioPlayer::begin(int, bool)
00:35:49.708 -> [D] AudioStreams.h : 51 - virtual void audio_tools::AudioStream::setAudioInfo(audio_tools::AudioInfo)
00:35:49.742 -> [I] AudioTypes.h : 104 - sample_rate: 44100
00:35:49.742 -> [I] AudioTypes.h : 105 - channels: 2
00:35:49.742 -> [I] AudioTypes.h : 106 - bits_per_sample: 16
00:35:49.742 -> [D] AudioEncoded.h : 329 - virtual bool audio_tools::EncodedAudioOutput::begin()
00:35:49.742 -> [D] AudioEncoded.h : 331 - virtual bool audio_tools::EncodedAudioOutput::begin()
00:35:49.742 -> [D] AudioEncoded.h : 28 - virtual void audio_tools::AudioDecoder::setAudioInfo(audio_tools::AudioInfo)
00:35:49.777 -> [D] CodecMP3Helix.h : 77 - virtual void audio_tools::MP3DecoderHelix::begin()
00:35:49.777 -> [D] MetaDataFilter.h : 30 - void audio_tools::MetaDataFilter<Decoder>::begin() [with Decoder = libhelix::MP3DecoderHelix]
00:35:49.777 -> [D] AudioSourceSDFAT.h : 75 - virtual void audio_tools::AudioSourceSDFAT::begin()
00:35:49.777 -> [D] SDDirect.h : 26 - void audio_tools::SDDirect<SDT, FileT>::begin(const char*, const char*, const char*) [with SDT = SdFat32; FileT = File32]
00:35:49.810 -> [I] MetaDataID3.h : 569 - virtual void audio_tools::MetaDataID3::begin()
00:35:49.810 -> [I] AudioSourceSDFAT.h : 101 - selectStream SDFAT: 0
00:35:49.810 -> [D] SDDirect.h : 76 - listDir: /
00:35:49.810 -> [D] SDDirect.h : 241 - FileT audio_tools::SDDirect<SDT, FileT>::open(const char*) [with SDT = SdFat32; FileT = File32]
00:35:49.810 -> [D] SDDirect.h : 143 - isDirectory /: 1
00:35:49.844 -> [D] SDDirect.h : 143 - isDirectory /: 1
00:35:49.844 -> [D] SDDirect.h : 128 - void audio_tools::SDDirect<SDT, FileT>::rewind(FileT) [with SDT = SdFat32; FileT = File32]
00:35:49.844 -> [D] SDDirect.h : 148 - FileT audio_tools::SDDirect<SDT, FileT>::openNext(FileT&) [with SDT = SdFat32; FileT = File32]
00:35:49.844 -> [D] SDDirect.h : 143 - isDirectory .Spotlight-V100: 1
00:35:49.844 -> [D] SDDirect.h : 211 - -> fileNamePath: .Spotlight-V100
00:35:49.877 -> [D] SDDirect.h : 224 - <- fileNamePath: /.Spotlight-V100
00:35:49.877 -> [D] SDDirect.h : 101 - name: /.Spotlight-V100
00:35:49.877 -> [D] SDDirect.h : 161 - void audio_tools::SDDirect<SDT, FileT>::pushPath(const char*) [with SDT = SdFat32; FileT = File32]

0:35:53.578 -> [D] SDDirect.h : 241 - FileT audio_tools::SDDirect<SDT, FileT>::open(const char*) [with SDT = SdFat32; FileT = File32]
00:35:53.578 -> [D] SDDirect.h : 143 - isDirectory journals.assisted_import_pre: 1
00:35:53.578 -> [D] SDDirect.h : 143 - isDirectory journals.assisted_import_pre: 1
00:35:53.578 -> [D] SDDirect.h : 128 - void audio_tools::SDDirect<SDT, FileT>::rewind(FileT) [with SDT = SdFat32; FileT = File32]
00:35:53.612 -> [D] SDDirect.h : 148 - FileT audio_tools::SDDirect<SDT, FileT>::openNext(FileT&) [with SDT = SdFat32; FileT = File32]
00:35:53.612 -> [D] SDDirect.h : 152 - No next file
00:35:53.612 -> [D] SDDirect.h : 168 - void audio_tools::SDDirect<SDT, FileT>::popPath() [with SDT = SdFat32; FileT = File32]
00:35:53.612 -> [D] SDDirect.h : 171 - popPath: journals.assisted_import_pre
00:35:53.646 -> [D] SDDirect.h : 148 - FileT audio_tools::SDDirect<SDT, FileT>::openNext(FileT&) [with SDT = SdFat32; FileT = File32]
00:35:53.646 -> [D] SDDirect.h : 143 - isDirectory journals.assisted_import_post: 1
00:35:53.646 -> [D] SDDirect.h : 211 - -> fileNamePath: journals.assisted_import_post
00:35:53.646 -> [D] SDDirect.h : 224 - <- fileNamePath: /.Spotlight-V100/Store-V2/8FE1E423-7918-4AFE-A9C0-0EA8EA13A718/journals.assisted_import_post
00:35:53.678 -> [D] SDDirect.h : 101 - name: /.Spotlight-V100/Store-V2/8FE1E423-7918-4AFE-A9C0-0EA8EA13A718/journals.assisted_import_post
00:35:53.678 -> [D] SDDirect.h : 161 - void audio_tools::SDDirect<SDT, FileT>::pushPath(const char*) [with SDT = SdFat32; FileT = File32]
00:35:53.678 -> [D] SDDirect.h : 162 - pushPath: journals.assisted_import_post
00:35:53.678 -> [D] SDDirect.h : 76 - listDir: /.Spotlight-V100/Store-V2/8FE1E423-7918-4AFE-A9C0-0EA8EA13A718/journals.assisted_import_post
00:35:53.714 -> [D] SDDirect.h : 241 - FileT audio_tools::SDDirect<SDT, FileT>::open(const char*) [with SDT = SdFat32; FileT = File32]
00:35:53.714 -> [D] AudioA2DP.h : 317 - a2dp_stream_source_sound_data: 512 -> 512
```

Further debugging on BLE, I was able to find that BLE host gets reset when I transition from BT to BLE, and stops responding.
The logs below showcase ESP32 finds the BLE beacon and see if it within 1 meter range. If yes it stops scanning, then it begin BT in slave mode. On pairing ESP32 with my phone, I sent a string "hello" to ESP32. On receiving, ESP32 should resume the scan again But what I see it that BLE host gets reset and after that I get GAP init failed error.
```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13192
load:0x40080400,len:3028
entry 0x400805e4
E (599) esp_core_dump_flash: No core dump  …ÉÑ¥Ñ¥½¹found!
E (599) esp_core_dump_flash: No core dump partition found!
firmware version 0.4
1
Scanning for peripherals
Advertised Device found: Name: , Address: 57:04:4b:a3:24:6b, manufacturer data: 4c001005311cb7c9f9, txPower: 12
Advertised Device found: Name: , Address: 5a:48:e8:ea:a7:b1, manufacturer data: 060001092002cc4e59b1f9d7c6284db5f38bca3482078467229cbc024e
Advertised Device found: Name: , Address: 49:ca:64:9f:8d:3c, manufacturer data: 4c0010021300, txPower: 12
Advertised Device found: Name: , Address: e7:be:62:3a:61:dc, manufacturer data: 4c0012020000
Found Beacon
-82
@1m
Scan Ended
HERE !!
BT init success !!
hello
CRIT NimBLEDevice: Resetting state; reason=19, 
CRIT NimBLEScan: Unable to scan - Host Reset
0
Scanning for peripherals
E NimBLEScan: Error initiating GAP discovery procedure; rc=21, 
0
```

Any clues?

Thanks

**Reply by pschatzmann** [reply]
Did you double check that the NimBLE initialization is not calling esp_bt_controller_enable(ESP_BT_MODE_BLE) ?
Are you sure that NimBLE is supporting this scenario ?

**Reply by cyberFIRM** [reply]
Yes I commented that in the library and esp_bt_controller_enable(ESP_BT_MODE_BTDM) is being called and used once. 
I mean the NimBLE should support this scenario, atleast theoretically it should.

I will discuss this with @h2zero too if I get a response from him.

Thanks

**Comment by pschatzmann** [comment]
I added a [simple example](https://github.com/pschatzmann/ESP32-A2DP/blob/main/examples/bt_music_receiver_and_BLE/bt_music_receiver_and_BLE.ino) for an A2DP Sink where the title is made available via BLE using the ESP32 Arduino BLE API

I can confirm that the same sketch with NimBLE is crashing and you will need to rewrite/correct the NimBLE::init() method

**Reply by cyberFIRM** [reply]
Hi @pschatzmann ,

Thanks for the great help. I was able to implement the application. One last thing, next step is to play multiple audio files.
Example:

Now I have 3 beacons and 3 audio mp3 files in my SD card. If ESP32 is within 20 cm of first beacon, it should play 1.mp3
If ESP32 is within 20cm of second beacon it should play 2.mp3 on BT speaker. Vice versa

What are the necessary steps required to achieve this? I searched a lot in your examples but the player.next() does not seems to work. It plays audio on the begin but if other beacon is found and I jump to the index of the associated mp3 it does not play. Any clues?

Do I need player.stop() and player.play() function before playing second audio?

**Reply by pschatzmann** [reply]
I can't tell you what you are doing wrong, but my guess would be that you don't debounce the request (and call it many times). Here is a [working example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd-audiokit/player-sd-audiokit.ino) that uses navigation. 

Just use the logger to figure out what's going on...

**Reply by cyberFIRM** [reply]
Hi @pschatzmann 

A quick question:
If you were to send 3 mp3's of duration 30 seconds in a loop to a bluetooth speaker from SD card, how would you do that?

Thanks

**Comment by pschatzmann** [comment]
I am not sure what you mean. If you want a better control over what is played when, don't use the AudioPlayer or if you want to stick to it use at least your own AudioSource
